src/core/ddsc/src/dds_entity.c

```diff
@@ -80,11 +80,34 @@ const ddsrt_avl_treedef_t dds_entity_children_td = DDSRT_AVL_TREEDEF_INITIALIZER
 static void dds_entity_observers_signal (dds_entity *observed, uint32_t status);
 static void dds_entity_observers_signal_delete (dds_entity *observed);
 
+static dds_return_t dds_delete_impl (dds_entity_t entity, enum delete_impl_state delstate);
+static dds_return_t really_delete_pinned_closed_locked (struct dds_entity *e, enum delete_impl_state delstate);
+
 void dds_entity_add_ref_locked (dds_entity *e)
 {
   dds_handle_add_ref (&e->m_hdllink);
 }
 
+void dds_entity_drop_ref (dds_entity *e)
+{
+  if (dds_handle_drop_ref (&e->m_hdllink))
+  {
+    dds_return_t ret = dds_delete_impl(e->m_hdllink.hdl, DIS_EXPLICIT);
```

Comment:
There was (some) method behind the madness of my "WiP" commit in setting the ``CLOSING`` flag on the handle in ``dds_handle_drop_ref`` (and ``dds_handle_unpin_and_drop_ref``). I'm no way trying to defend the body of this function (and the next) in that commit, that was simply an ugly hack along the way.

The way you do it here looks perfectly fine, but what it does allow is for some thread to increment the refcount after it dropped to 0 but before ``dds_delete_impl`` is called, if there is a path that allows another thread to do so. One has to wonder whether that's an actual problem, and there are two perspectives on it:

1. Is there any way in which that can happen? — If the answer is "no", then all is well.
2. If it happens, does it behave in an acceptable manner? — If it always behaves in an acceptable manner (not crashing for example) then all is well, too.

Re 1: for ``dds_entity_drop_ref``, it really is only a topic that can be passed in, and it would already have to be in the "deferred delete" state (otherwise the implied reference from the handle will keep it alive). I'm pretty sure it is possible, but it can be prevented by checking in ``dds_create_reader`` and ``dds_create_writer``. That is a manageable case.

``dds_entity_unpin_and_drop_ref`` is only used for the global "cyclone" entity and the domain entities. Those are also really special cases, sneak around the regular destruction path and have additional locking to prevent such games. I just haven't figured that one out yet ...

Re 2: For topics, it could cause the call to ``delete`` to return ``ALREADY_DELETED``, but I think nothing worse than that. One could change the assert, one could even decide that ``ALREADY_DELETED`` doesn't make sense and that it should always return ``OK`` anyway.

For the domain and cyclonedds objects, I think it is equivalent to simply calling ``dds_delete`` from the application code. It may therefore result in deleting a domain/participant that was being created in parallel, but I do think that afterward that new one either exists and functions normally, or doesn't exist at all. Given that it is somewhat daft use of the API, I'd say that is acceptable.

In short, your simplification looks reasonable; in this form it will work better than the code currently in master (e.g., calling ``dds_delete`` twice on a topic is deadly now); and the possibly problematic cases seem to be rather far out there in terms of realistic use.

It'd be interesting to try testing it a bit, for example by creating and deleting participants in some domain on multiple threads.

Reply:
I've added a torture test that showed that a wrong mutex was used when adding a participant to the domain as a child.

Reply:
It still crashes intermittently when compiling with gcc.
Somehow, I can't reproduce it with clang.

Reply:
There are two torture tests. One triggers an exiting issue (the crash) that is somewhat outside the scope of this PR and is related to actually init/deinit the library. For that reason, I've disabled that torture test.